modules/auction_manager/src/lib.rs

```diff
@@ -734,134 +723,138 @@ impl<T: Trait> Module<T> {
-		if let Some(collateral_auction) = Self::collateral_auctions(auction_id) {
-			if let Some((bidder, bid_price)) = winner {
-				// decrease account ref of bidder
-				system::Module::<T>::dec_ref(&bidder);
-
-				let mut should_deal = true;
-
-				// if bid_price doesn't reach target and trading with DEX will get better result
-				if !collateral_auction.in_reverse_stage(bid_price)
-					&& bid_price
-						< T::DEX::get_target_amount(
-							collateral_auction.currency_id,
-							T::GetStableCurrencyId::get(),
-							collateral_auction.amount,
-						) {
-					// try trade with DEX
-					if let Ok(amount) = T::CDPTreasury::swap_collateral_to_stable(
-						collateral_auction.currency_id,
-						collateral_auction.amount,
-						Zero::zero(),
-					) {
-						// swap successfully, will not deal
-						should_deal = false;
-
-						// refund stable currency to the last bidder, ignore result to continue
-						let _ = T::CDPTreasury::issue_debit(&bidder, bid_price, false);
-
-						if collateral_auction.in_reverse_stage(amount) {
-							// refund extra stable currency to recipient, ignore result to continue
-							let _ = T::CDPTreasury::issue_debit(
-								&collateral_auction.refund_recipient,
-								amount
-									.checked_sub(collateral_auction.target)
-									.expect("ensured amount > target; qed"),
-								false,
-							);
-						}
+	fn collateral_auction_end_handler(
+		auction_id: AuctionId,
+		collateral_auction: CollateralAuctionItem<T::AccountId, T::BlockNumber>,
+		winner: Option<(T::AccountId, Balance)>,
+	) {
+		if let Some((bidder, bid_price)) = winner {
+			// decrease account ref of bidder
+			system::Module::<T>::dec_ref(&bidder);
 
-						<Module<T>>::deposit_event(RawEvent::DEXTakeCollateralAuction(
-							auction_id,
-							collateral_auction.currency_id,
-							collateral_auction.amount,
-							amount,
-						));
-					}
-				}
+			let mut should_deal = true;
 
-				if should_deal {
-					// transfer collateral to winner from CDP treasury, ignore result to continue
-					let _ = T::CDPTreasury::withdraw_collateral(
-						&bidder,
+			// if bid_price doesn't reach target and trading with DEX will get better result
+			if !collateral_auction.in_reverse_stage(bid_price)
+				&& bid_price
+					< T::DEX::get_target_amount(
 						collateral_auction.currency_id,
+						T::GetStableCurrencyId::get(),
 						collateral_auction.amount,
+					) {
+				// try trade with DEX
+				if let Ok(amount) = T::CDPTreasury::swap_collateral_to_stable(
+					collateral_auction.currency_id,
+					collateral_auction.amount,
+					Zero::zero(),
+				) {
+					// swap successfully, will not deal
+					should_deal = false;
+
+					// refund stable currency to the last bidder, ignore result to continue
+					T::CDPTreasury::issue_debit(&bidder, bid_price, false).expect(
+						"issue_debit calls currency::deposit which always resolves so there's no error here; qed",
```

Comment:
this is still not a solid proof. what happen if the implementation changed in future?
should just ignore the error, and add a comment say it currently cannot fail. and it is ok to ignore the error in this case.

Reply:
Bryan means this https://github.com/AcalaNetwork/Acala/issues/366